File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/sccolor.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const ScColorLength = 32

type ScColor struct {
	id [ScColorLength]byte
}

var (
	IOTA = ScColor{}
	MINT = ScColor{}
)

func init() {
	for i := range MINT.id {
		MINT.id[i] = 0xff
	}
}

func (o ScColor) Bytes() []byte {
	return ColorToBytes(o)
}

func (o ScColor) String() string {
	return ColorToString(o)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func ColorDecode(dec *WasmDecoder) ScColor {
	return colorFromBytesUnchecked(dec.FixedBytes(ScColorLength))
}

func ColorEncode(enc *WasmEncoder, value ScColor) {
	enc.FixedBytes(value.Bytes(), ScColorLength)
}

func ColorFromBytes(buf []byte) ScColor {
	if len(buf) == 0 {
		return ScColor{}
	}
	if len(buf) != ScColorLength {
		panic("invalid Color length")
	}
	return colorFromBytesUnchecked(buf)
}

func ColorToBytes(value ScColor) []byte {
	return value.id[:]
}

func ColorToString(value ScColor) string {
	// TODO standardize human readable string
	return Base58Encode(value.id[:])
}

func colorFromBytesUnchecked(buf []byte) ScColor {
	o := ScColor{}
	copy(o.id[:], buf)
	return o
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableColor struct {
	proxy Proxy
}

func NewScImmutableColor(proxy Proxy) ScImmutableColor {
	return ScImmutableColor{proxy: proxy}
}

func (o ScImmutableColor) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableColor) String() string {
	return ColorToString(o.Value())
}

func (o ScImmutableColor) Value() ScColor {
	return ColorFromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableColor struct {
	ScImmutableColor
}

func NewScMutableColor(proxy Proxy) ScMutableColor {
	return ScMutableColor{ScImmutableColor{proxy: proxy}}
}

func (o ScMutableColor) Delete() {
	o.proxy.Delete()
}

func (o ScMutableColor) SetValue(value ScColor) {
	o.proxy.Set(ColorToBytes(value))
}

File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/sccolor.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import {base58Encode, WasmDecoder, WasmEncoder, zeroes} from "./codec";
import {Proxy} from "./proxy";
import {bytesCompare} from "./scbytes";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScColorLength = 32;

export class ScColor {
    id: u8[] = zeroes(ScColorLength);

    constructor(fill: u8) {
        this.id.fill(fill);
    }

    public equals(other: ScColor): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): u8[] {
        return colorToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        // TODO standardize human readable string
        return colorToString(this);
    }
}

// predefined colors
export const IOTA: ScColor = new ScColor(0x00);
export const MINT: ScColor = new ScColor(0xff);

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function colorDecode(dec: WasmDecoder): ScColor {
    return colorFromBytesUnchecked(dec.fixedBytes(ScColorLength));
}

export function colorEncode(enc: WasmEncoder, value: ScColor): void {
    enc.fixedBytes(value.toBytes(), ScColorLength);
}

export function colorFromBytes(buf: u8[]): ScColor {
    if (buf.length == 0) {
        return new ScColor(0);
    }
    if (buf.length != ScColorLength) {
        panic("invalid Color length");
    }
    return colorFromBytesUnchecked(buf);
}

export function colorToBytes(value: ScColor): u8[] {
    return value.id;
}

export function colorToString(value: ScColor): string {
    // TODO standardize human readable string
    return base58Encode(value.id);
}

function colorFromBytesUnchecked(buf: u8[]): ScColor {
    let o = new ScColor(0);
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableColor {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return colorToString(this.value());
    }

    value(): ScColor {
        return colorFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableColor extends ScImmutableColor {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScColor): void {
        this.proxy.set(colorToBytes(value));
    }
}
